packages/balances/src/util/hydrateChaindata.ts

```diff
@@ -0,0 +1,47 @@
+import { EvmTokenFetcher, MiniMetadataUpdater } from "@talismn/balances"
+import { ChaindataProvider } from "@talismn/chaindata-provider"
+
+/** Pulls the latest chaindata from https://github.com/TalismanSociety/chaindata */
+export const hydrateChaindataAndMiniMetadata = async (
+  chaindataProvider: ChaindataProvider,
+  miniMetadataUpdater: MiniMetadataUpdater
+) => {
+  await Promise.all([
+    miniMetadataUpdater.hydrateFromChaindata(),
+    miniMetadataUpdater.hydrateCustomChains(),
+    chaindataProvider.hydrateChains(),
+    chaindataProvider.hydrateEvmNetworks(),
+  ])
+
+  const chains = await chaindataProvider.chains()
+  const { statusesByChain } = await miniMetadataUpdater.statuses(chains)
+  const goodChains = [...statusesByChain.entries()].flatMap(([chainId, status]) =>
+    status === "good" ? chainId : []
+  )
+  await chaindataProvider.hydrateTokens(goodChains)
+}
+
+/** Builds any missing miniMetadatas (e.g. for the user's custom substrate chains) */
+export const updateCustomMiniMetadata = async (
+  chaindataProvider: ChaindataProvider,
+  miniMetadataUpdater: MiniMetadataUpdater,
+  isTest?: boolean
+) => {
+  // Don't update custom minimetadata in tests
+  //
+  // TODO: Remove this, and instead mock the websocket response for all of the called rpc methods.
+  // E.g. state_getMetadata, system_properties, etc
+  if (isTest) return
```

Comment:
Checking for test state in the code itself is definitely not a good way to do things, so I would really prefer not to elevate that into a function argument. If we can't mock it, can we stick with the existing constant check?

Reply:
The problem with that is that the `TEST` constant comes from the extension package, which depends on the `@talismn/balances` package where this code is - not the other way around.